File: alumnium/server/api_models.py
from typing import Any, List, Optional

from pydantic import BaseModel, Field

from .agents.retriever_agent import Data


# Base versioned model
class VersionedModel(BaseModel):
    api_version: str = Field(default="v1", description="API version")


class SessionRequest(VersionedModel):
    provider: str
    name: Optional[str] = None
    tools: List[dict[str, Any]]
    platform: str  # chromium, xcuitest, or uiautomator2


class SessionResponse(VersionedModel):
    session_id: str


class PlanRequest(VersionedModel):
    goal: str
    accessibility_tree: str
    url: Optional[str] = None
    title: Optional[str] = None
    area_id: Optional[int] = None


class PlanResponse(VersionedModel):
    steps: List[str]


class StepRequest(VersionedModel):
    goal: str
    step: str
    accessibility_tree: str
    area_id: Optional[int] = None


class StepResponse(VersionedModel):
    actions: List[dict[str, Any]]


class StatementRequest(VersionedModel):
    statement: str
    accessibility_tree: str
    url: Optional[str] = None
    title: Optional[str] = None
    screenshot: Optional[str] = None  # base64 encoded image
    area_id: Optional[int] = None


class StatementResponse(VersionedModel):
    # TODO: Move typecasting to the client
    result: Data
    explanation: str


class AreaRequest(VersionedModel):
    description: str
    accessibility_tree: str


class AreaResponse(VersionedModel):
    id: int
    explanation: str


class FindRequest(VersionedModel):
    description: str
    accessibility_tree: str
    area_id: Optional[int] = None


class FindResponse(VersionedModel):
    elements: list[dict[str, int | str]]


class AddExampleRequest(VersionedModel):
    goal: str
    actions: List[str]


class AddExampleResponse(VersionedModel):
    success: bool
    message: str


class ClearExamplesResponse(VersionedModel):
    success: bool
    message: str


class CacheResponse(VersionedModel):
    success: bool
    message: str


class ErrorResponse(VersionedModel):
    error: str
    detail: Optional[str] = None

File: packages/core/src/clients/ApiModels.ts
/**
 * TypeScript interfaces for API request and response models.
 *
 * IMPORTANT: These interfaces must be kept in sync with the Python API models
 * defined in packages/python/src/alumnium/server/api_models.py
 */

import type { ElementRef } from "../server/serverSchema.js";
import type { ToolCall } from "../tools/BaseTool.js";

export interface SessionRequest {
  platform: "chromium" | "uiautomator2" | "xcuitest";
  provider: string;
  name?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  tools: { [key: string]: any }[];
  planner: boolean;
  excluded_attributes?: string[];
}

export interface SessionResponse {
  session_id: string;
}

export interface PlanRequest {
  goal: string;
  accessibility_tree: string;
  url?: string;
  title?: string;
  app?: string;
}

export interface PlanResponse {
  explanation: string;
  steps: string[];
}

export interface StepRequest {
  goal: string;
  step: string;
  accessibility_tree: string;
  app?: string;
}

export interface StepResponse {
  explanation: string;
  actions: ToolCall[];
}

export interface StatementRequest {
  statement: string;
  accessibility_tree: string;
  url?: string;
  title?: string;
  screenshot?: string | null;
  app?: string;
}

export interface StatementResponse {
  result: string | string[];
  explanation: string;
}

export interface AreaRequest {
  description: string;
  accessibility_tree: string;
  app?: string;
}

export interface AreaResponse {
  id: number;
  explanation: string;
}

export interface FindRequest {
  description: string;
  accessibility_tree: string;
  app?: string;
}

export interface FindResponse {
  elements: ElementRef[];
}

export interface AddExampleRequest {
  goal: string;
  actions: string[];
}

export interface ChangeState {
  accessibility_tree: string;
  url: string;
}

export interface ChangesRequest {
  before: ChangeState;
  after: ChangeState;
}

export interface ChangesResponse {
  result: string;
}

export interface StatsResponse {
  [key: string]: {
    [key: string]: number;
  };
}
